### PR TITLE
Make JIT and pointwise work with zero input args

### PR DESCRIPTION
**Pull request by TedThemistokleous**
Ensure that we don't have empty inputs when computing shape for pointwise function

### CODE REVIEW

File: src/include/migraphx/op/pointwise.hpp [hunk 1]
```diff
@@ -41,6 +41,8 @@ struct pointwise
 
     shape compute_shape(const std::vector<shape>& inputs, std::vector<module_ref> mods) const
     {
+        assert(inputs.size() > 0);
```
Comment (umangyadav) [inline src/include/migraphx/op/pointwise.hpp:44]: this should be using `check_shapes().has_atleast(1)`.

**Comment by umangyadav** [comment]
Add some tests in `op_shape_tests.cpp` as well for pointwise.

**Comment by umangyadav** [comment]
You can also modify `fuse_pointwise` with this change so that it doesn't attempt to create pointwise op with empty input args.

**Comment by TedThemistokleous** [comment]
> You can also modify `fuse_pointwise` with this change so that it doesn't attempt to create pointwise op with empty input args.

This is probably a good idea as well

> Add some tests in `op_shape_tests.cpp` as well for pointwise.

Will do

File: src/include/migraphx/op/pointwise.hpp (hunk 1 above)
Comment (TedThemistokleous) [reply src/include/migraphx/op/pointwise.hpp:44]: On it

Comment (TedThemistokleous) [reply src/include/migraphx/op/pointwise.hpp]: done

File: src/include/migraphx/op/pointwise.hpp [hunk 2]
```diff
@@ -41,6 +41,8 @@ struct pointwise
 
     shape compute_shape(const std::vector<shape>& inputs, std::vector<module_ref> mods) const
     {
+        check_shapes(inputs, *this, true).has_at_least(1);
```
Comment (umangyadav) [inline src/include/migraphx/op/pointwise.hpp:44]:
```suggestion
        check_shapes(inputs, *this).has_at_least(1);
```
`true` is for dynamic shapes.

Comment (TedThemistokleous) [reply src/include/migraphx/op/pointwise.hpp:44]: Oh, I thought it was inclusive of both static and dynamic outcomes. I'll fix

File: src/fuse_pointwise.cpp [hunk 3]
```diff
@@ -67,6 +67,8 @@ static void create_pointwise_modules(module_pass_manager& mpm)
             continue;
         if(ins->get_operator().name() == "layout")
             continue;
+        if(ins->inputs().size() == 0)
+            continue;
```
Comment (umangyadav) [inline src/fuse_pointwise.cpp:71]: This wouldn't be necessary

**Comment by umangyadav** [comment]
This requires unit/verify tests.

File: src/fuse_pointwise.cpp (hunk 3 above)
Comment (TedThemistokleous) [reply src/fuse_pointwise.cpp]: Removed

**Comment by TedThemistokleous** [comment]
related to #1587

File: test/verify/test_gather_literal_inputs.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        mm->add_instruction(migraphx::make_op("gather", {{"axis", axis}}), a0, a1);
+
+        mm->debug_print();
+
+        return p;
+    }
+};
+
+// Standard gather test
+template struct test_gather_literal_inputs<0>;
```
Comment (umangyadav) [inline test/verify/test_gather_literal_inputs.cpp:53]: No need to make it templated unless there are multiple cases.

File: test/verify/test_gather_literal_inputs.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        migraphx::shape s{migraphx::shape::int64_type, {3}};
+        migraphx::shape g_shape{migraphx::shape::int64_type, {1}, {0}};
+        migraphx::shape s_indices{migraphx::shape::int32_type, {3}};
+        std::vector<int> indices{3, 800, 800};
+        auto a0  = mm->add_literal(migraphx::literal{s_indices, indices});
+        auto a1  = mm->add_literal(migraphx::literal{g_shape, {1}});
+        int axis = Axis;
+        mm->add_instruction(migraphx::make_op("gather", {{"axis", axis}}), a0, a1);
+
+        mm->debug_print();
```
Comment (umangyadav) [inline test/verify/test_gather_literal_inputs.cpp:46]: Remove this

File: test/verify/test_gather_literal_inputs.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+#include <migraphx/make_op.hpp>
+
+template <int Axis>
+struct test_gather_literal_inputs : verify_program<test_gather_literal_inputs<Axis>>
+{
+    migraphx::program create_program() const
+    {
+        migraphx::program p;
+        auto* mm = p.get_main_module();
+        migraphx::shape s{migraphx::shape::int64_type, {3}};
```
Comment (umangyadav) [inline test/verify/test_gather_literal_inputs.cpp:37]: Is this `s` used anywhere ?

File: src/simplify_reshapes.cpp [hunk 7]
```diff
@@ -185,6 +185,25 @@ struct find_transpose
     }
 };
 
+struct find_literal_convert
```
Comment (umangyadav) [inline src/simplify_reshapes.cpp:188]: This is a different issue, it would be better to keep it as part of a separate PR

Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp:188]: All good. I'll remove then

**Comment by TedThemistokleous** [comment]
Removed matcher related optimization for now. for convert(literal)

File: test/verify/test_gather_literal_inputs.cpp (hunk 6 above)
Comment (TedThemistokleous) [reply test/verify/test_gather_literal_inputs.cpp:37]: No ill remove

File: test/verify/test_gather_literal_inputs.cpp (hunk 5 above)
Comment (TedThemistokleous) [reply test/verify/test_gather_literal_inputs.cpp:46]: Will do

File: test/op_shape_test.cpp [hunk 8]
```diff
@@ -1819,6 +1819,33 @@ TEST_CASE(pad_dyn_shape1)
     expect_shape(output, migraphx::make_op("pad", {{"pads", {0, 0, 1, 1, 0, 0, 1, 1}}}), input);
 }
 
+TEST_CASE(pointwise_no_module)
+{
+    migraphx::shape input{migraphx::shape::float_type, {0}, {0}};
+    throws_shape(migraphx::make_op("pointwise"), input);
+}
+
+TEST_CASE(pointwise_no_input)
```
Comment (umangyadav) [inline test/op_shape_test.cpp:1828]: what is this test is checking for ? It should have EXPECT or throws_shape.

File: test/op_shape_test.cpp [hunk 9]
```diff
@@ -1819,6 +1819,33 @@ TEST_CASE(pad_dyn_shape1)
...
+    auto* mm = p.get_main_module();
+    migraphx::module m;
+    std::vector<migraphx::instruction_ref> args{};
+    migraphx::shape output{migraphx::shape::float_type, {1}, {0}};
+    auto l = m.add_literal(migraphx::literal(output, {1}));
+    m.add_return({l});
+    mm->add_instruction(migraphx::make_op("pointwise"), args, {&m});
+}
+
+TEST_CASE(pointwise_no_output)
```
Comment (umangyadav) [inline test/op_shape_test.cpp:1840]:
```suggestion
TEST_CASE(pointwise_no_input_args)
```

File: test/op_shape_test.cpp (hunk 8 above)
Comment (TedThemistokleous) [reply test/op_shape_test.cpp:1828]: Added

File: test/op_shape_test.cpp (hunk 9 above)
Comment (TedThemistokleous) [reply test/op_shape_test.cpp:1841]: This is intentional so we dont have any output shape assigned for the module going into pointwise. since this block in pointwise

```
      if(pm->get_output_shapes().size() != 1)
            MIGRAPHX_THROW("submodule should have only one output.");
        // if pointwise op doens't have any input args (const-foldable) then just return output
        // shape
```

File: test/verify/test_gather_literal_inputs.cpp (hunk 4 above)
Comment (umangyadav) [reply test/verify/test_gather_literal_inputs.cpp:50]: We can add `Axis` if in the future if we need to add tests for multiple axes.

Comment (TedThemistokleous) [reply test/verify/test_gather_literal_inputs.cpp]: Sure. Removed.

**Review by umangyadav — APPROVED** [review]

File: src/include/migraphx/op/pointwise.hpp [hunk 10]
```diff
@@ -45,14 +45,19 @@ struct pointwise
...
-        auto* pm    = mods.front();
+        auto* pm = mods.front();
+        if(pm->get_output_shapes().size() != 1)
+            MIGRAPHX_THROW("submodule should have only one output.");
+        // if pointwise op doens't have any input args (const-foldable) then just return output
+        // shape
+        if(inputs.empty())
+        {
+            return pm->get_output_shapes().front();
+        }
```
Comment (pfultz2) [inline src/include/migraphx/op/pointwise.hpp:56]: This should throw an error.

File: src/include/migraphx/op/pointwise.hpp [hunk 11]
```diff
@@ -45,14 +45,19 @@ struct pointwise
         {
             MIGRAPHX_THROW("should have one submodule.");
         }
-        auto* pm    = mods.front();
+        auto* pm = mods.front();
+        if(pm->get_output_shapes().size() != 1)
+            MIGRAPHX_THROW("submodule should have only one output.");
+        // if pointwise op doens't have any input args (const-foldable) then just return output
```
Comment (pfultz2) [inline src/include/migraphx/op/pointwise.hpp:51]: We shouldn't be creating a pointwise op in `fuse_pointwise`. Ideally, there should be a check for this.

File: src/cpp_generator.cpp [hunk 12]
```diff
@@ -218,6 +218,8 @@ std::string cpp_generator::create_function(const cpp_generator::function& f)
     std::string name = f.name.empty() ? "f" + std::to_string(impl->function_count) : f.name;
     impl->fs << join_strings(f.attributes, " ") << " " << f.return_type << " " << name;
     char delim = '(';
+    impl->fs << delim;
+    delim = ' '; // there should not be any delimiter for first input arg
```
Comment (pfultz2) [inline src/cpp_generator.cpp:222]: I would rather do:

```cpp
if (f.params.empty())
    impl->fs << delim;
for(auto&& p : f.params)
{
    ...
}
```

Comment (TedThemistokleous) [reply src/cpp_generator.cpp]: Sure. Done

File: src/include/migraphx/op/pointwise.hpp [hunk 13]
```diff
@@ -45,14 +45,15 @@ struct pointwise
         {
             MIGRAPHX_THROW("should have one submodule.");
         }
-        auto* pm    = mods.front();
+        auto* pm = mods.front();
+        if(pm->get_output_shapes().size() != 1)
+            MIGRAPHX_THROW("submodule should have only one output.");
+        if(inputs.empty())
+            MIGRAPHX_THROW("submodule should have atleast one input");
```
Comment (pfultz2) [inline src/include/migraphx/op/pointwise.hpp:52]: This should probably say `pointwise should have at least one input`.

Comment (TedThemistokleous) [reply src/include/migraphx/op/pointwise.hpp]: Done

File: src/include/migraphx/op/pointwise.hpp (hunk 10 above)
Comment (TedThemistokleous) [reply src/include/migraphx/op/pointwise.hpp]: Done

**Review by pfultz2 — APPROVED** [review]

File: test/verify/test_gather_literal_inputs.cpp [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        migraphx::shape s_indices{migraphx::shape::int32_type, {3}};
+        std::vector<int> indices{3, 800, 800};
+        auto a0  = mm->add_literal(migraphx::literal{s_indices, indices});
+        auto a1  = mm->add_literal(migraphx::literal{g_shape, {1}});
+        int axis = 0;
+        mm->add_instruction(migraphx::make_op("gather", {{"axis", axis}}), a0, a1);
+
+        return p;
+    }
+};
```
Comment (pfultz2) [inline test/verify/test_gather_literal_inputs.cpp:46]: This test should be moved to fuse_pointwise.cpp. Also fuse_pointwise needs to be updated to avoid the nullary pointwise op.

**Review by pfultz2 — CHANGES_REQUESTED** [review]
fuse_pointwise needs to be updated.

File: test/verify/test_gather_literal_inputs.cpp (hunk 14 above)
Comment (pfultz2) [reply test/verify/test_gather_literal_inputs.cpp:46]: Oops this is not the test that is throwing.

File: test/op_shape_test.cpp [hunk 15]
```diff
@@ -1819,6 +1819,34 @@ TEST_CASE(pad_dyn_shape1)
     expect_shape(output, migraphx::make_op("pad", {{"pads", {0, 0, 1, 1, 0, 0, 1, 1}}}), input);
 }
 
+TEST_CASE(pointwise_no_module)
+{
+    migraphx::shape input{migraphx::shape::float_type, {0}, {0}};
+    throws_shape(migraphx::make_op("pointwise"), input);
+}
+
+TEST_CASE(pointwise_no_input)
```
[the remark this answers is not in the document]
Comment (TedThemistokleous) [reply test/op_shape_test.cpp:1828]: Done

File: test/verify/test_gather_literal_inputs.cpp (hunk 14 above)
Comment (TedThemistokleous) [reply test/verify/test_gather_literal_inputs.cpp:46]: Should I keep this in then and not move it to fuse_pointwise?

File: test/fuse_pointwise.cpp [hunk 16]
```diff
@@ -329,4 +329,21 @@ TEST_CASE(all_scalar_input)
...
+        migraphx::shape g_shape{migraphx::shape::int64_type, {1}, {0}};
+        migraphx::shape s_indices{migraphx::shape::int32_type, {3}};
+        std::vector<int> indices{3, 800, 800};
+        auto a0  = mm->add_literal(migraphx::literal{s_indices, indices});
+        auto a1  = mm->add_literal(migraphx::literal{g_shape, {1}});
+        int axis = 0;
+        auto out = mm->add_instruction(migraphx::make_op("gather", {{"axis", axis}}), a0, a1);
+        mm->add_return({out});
+    }
+    EXPECT(test::throws([&] {run_pass(p); }));
```
Comment (pfultz2) [inline test/fuse_pointwise.cpp:346]: The pass shouldn't throw.

Comment (TedThemistokleous) [reply test/fuse_pointwise.cpp]: done

**Comment by TedThemistokleous** [comment]
> fuse_pointwise needs to be updated.

Done

**Review by pfultz2 — APPROVED** [review]
